### PR TITLE
Makes user mutable with the option to make immutable

### PR DESCRIPTION
**Pull request by ttomsu**
@ajordens @cfieber @jtk54 PTAL

I'll merge this once the other dependent CLs have also been LGTM'ed

### CODE REVIEW

File: kork-security/src/test/groovy/com/netflix/spinnaker/security/UserSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    then:
+      user.email == "batman"
+      user.allowedAccounts == ["abc", "def"]
+  }
+
+  def "should NOT allow modification when immutable"() {
+    setup:
+      def accounts = ["abc"]
+      def user = new User(email: "email", allowedAccounts: accounts).asImmutable()
```
Comment (cfieber) [inline kork-security/src/test/groovy/com/netflix/spinnaker/security/UserSpec.groovy]: could make this:
````groovy
def baseUser = new User(email: "email", allowedAccounts: accounts)
def user = baseUser.asImmutable()
````

And add an assertion that a mutation on `baseUser` isn't reflected on `user`

**Comment by ttomsu** [comment]
@duftler FYI

File: kork-security/src/main/groovy/com/netflix/spinnaker/security/User.groovy [hunk 2]
```diff
@@ -56,4 +60,17 @@ class User implements Serializable, UserDetails {
...
   boolean isEnabled() { return true }
+
+  @Immutable
+  class ImmutableUser extends User {
+    String email = User.this.email
+    String username = User.this.username
+    String firstName = User.this.firstName
+    String lastName = User.this.lastName
+
+    // Use .collect() to use snapshot a copy of these collections.
```
Comment (duftler) [inline kork-security/src/main/groovy/com/netflix/spinnaker/security/User.groovy]: Comment reads a bit funny.

Comment (ttomsu) [inline kork-security/src/main/groovy/com/netflix/spinnaker/security/User.groovy]: Fixed.

**Comment by duftler** [comment]
LGTM.

File: kork-security/src/test/groovy/com/netflix/spinnaker/security/UserSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  def "should not reflect changes to the mutable user in the immutable user"() {
+    setup:
+      def accounts = ["abc"]
+      def mutableUser = new User(email: "email", allowedAccounts: accounts)
+      def immutableUser = mutableUser.asImmutable()
+
+    expect:
+      mutableUser.email == "email"
+      mutableUser.allowedAccounts == ["abc"]
+      immutableUser.email == "email"
```
Comment (jtk54) [inline kork-security/src/test/groovy/com/netflix/spinnaker/security/UserSpec.groovy]: I would add a check here for `username == "email"` here too to check the wiring also.

**Comment by jtk54** [comment]
LGTM after comments.

File: kork-security/src/test/groovy/com/netflix/spinnaker/security/UserSpec.groovy (hunk 3 above)
Comment (ttomsu) [inline kork-security/src/test/groovy/com/netflix/spinnaker/security/UserSpec.groovy]: It appears that this is not happening. Nice find!

**Comment by cfieber** [comment]
LGTM

**Pull request merged**